src/Microsoft.ML.CpuMath/SseIntrinsics.cs

```diff
@@ -285,54 +285,164 @@ public static unsafe void MatMul(float[] mat, float[] src, float[] dst, int crow
+                            x2 = Sse.Multiply(x2, x1);
+                            result = Sse.Add(result, x2);
+                            ppos++;
+                        }
+
+                        Sse.Store(pDstCurrent, result);
+                        pDstCurrent += 4;
+                        pm0 += 4 * ccol;
+                    }
+                }
+                else
+                {
+                    if (misalignment != 0)
+                    {
+                        misalignment >>= 2;
+                        misalignment = 4 - misalignment;
+
+                        Vector128<float> mask = Sse.LoadVector128(((float*)(pLeadingAlignmentMask)) + (misalignment * 4));
 
-                    int* ppos = pposMin;
+                        float* pm1 = pm0 + ccol;
+                        float* pm2 = pm1 + ccol;
+                        float* pm3 = pm2 + ccol;
+                        Vector128<float> result = Sse.SetZeroVector128();
 
-                    while (ppos < pposEnd)
+                        int* ppos = pposMin;
+
+                        while (ppos < pposEnd)
+                        {
+                            int col = *ppos;
+                            Vector128<float> x1 = Sse.SetVector128(pm3[col], pm2[col], pm1[col], pm0[col]);
+
+                            x1 = Sse.And(mask, x1);
+                            Vector128<float> x2 = Sse.SetAllVector128(pSrcCurrent[col]);
+                            x2 = Sse.Multiply(x2, x1);
+                            result = Sse.Add(result, x2);
+                            ppos++;
+                        }
+
+                        Sse.Store(pDstCurrent, result);
+                        pDstCurrent += misalignment;
+                        pm0 += misalignment * ccol;
+                        length -= misalignment;
+                    }
+
+                    if (length > 3)
                     {
-                        int col = *ppos;
-                        Vector128<float> x1 = Sse.SetVector128(pm3[col], pm2[col], pm1[col], pm0[col]);
-                        Vector128<float> x2 = Sse.SetAllVector128(pSrcCurrent[col]);
-                        x2 = Sse.Multiply(x2, x1);
-                        result = Sse.Add(result, x2);
+                        remainder = length % 4;
+                        while (pDstCurrent < pDstEnd)
+                        {
+                            float* pm1 = pm0 + ccol;
+                            float* pm2 = pm1 + ccol;
+                            float* pm3 = pm2 + ccol;
+                            Vector128<float> result = Sse.SetZeroVector128();
+
+                            int* ppos = pposMin;
 
-                        ppos++;
+                            while (ppos < pposEnd)
+                            {
+                                int col = *ppos;
+                                Vector128<float> x1 = Sse.SetVector128(pm3[col], pm2[col], pm1[col], pm0[col]);
```

Comment:
What exactly about this algorithm needs alignment?  I don't see us calling any "Aligned" methods. Is `Sse.Store(pDstCurrent, result);` supposed to be `Sse.StoreAligned(pDstCurrent, result);`?

And if that is the only thing that needs alignment, do we know that this extra complexity outweighs the perf difference?

Reply:
Actually, looking at the algorithm, this isn't doing anything where the alignment should matter, since it is operating on sparse data (and has to do 4 float loads, rather than a single 16-byte load).

Reply:
The only one that I identified was the `Sse.Store(pDstCurrent, result)`.

Reply:
Ok, this makes sense to me. Thanks for the explanation.

Reply:
Thinking a bit more - do we want to change the other `MatMul` algorithm then? That one is aligning the input "matrix", and not the resulting destination.
Although, we are doing `(4 * ccol)` aligned loads for every 1 store. So maybe it is best to align that many loads?

Reply:
> The only one that I identified was the Sse.Store(pDstCurrent, result)

yes earlier it was doing one alignedLoad(it was removed while removing the add flag) and alignedStore.

Reply:
> Although, we are doing (4 * ccol) aligned loads for every 1 store. So maybe it is best to align that many loads?

yes that was motivation behind the aligning the input matrix because it has larger number of loads.